### PR TITLE
update publicips service to use Azure SDK v2

### LINKED ISSUE #3919: Update publicips service to SDK v2

### PR DESCRIPTION
**Pull request by nawazkh**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
- This PR will update publicips service to use Azure SDK v2

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #3919 

**Special notes for your reviewer**:

- [ ] cherry-pick candidate

**TODOs**:

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:

```release-note
update publicips service to use Azure SDK v2
```

### CODE REVIEW

**Comment by nawazkh** [comment]
Current PR is dependent on https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/3969 so I will be cherry-picking the changes from https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/3969 onto current PR to resolve assignment error at https://github.com/nawazkh/cluster-api-provider-azure/blob/c20099b2b691143f1034ef8b9309ed932c8e1b7e/azure/services/loadbalancers/spec.go#L146

File: azure/services/publicips/spec.go [hunk 1]
```diff
@@ -59,44 +59,44 @@ func (s *PublicIPSpec) OwnerResourceName() string {
...
-			PublicIPAddressVersion:   addressVersion,
-			PublicIPAllocationMethod: network.IPAllocationMethodStatic,
+		Properties: &armnetwork.PublicIPAddressPropertiesFormat{
+			PublicIPAddressVersion:   &addressVersion,
+			PublicIPAllocationMethod: ptr.To(armnetwork.IPAllocationMethodStatic),
 			DNSSettings:              dnsSettings,
 			IPTags:                   converters.IPTagsToSDK(s.IPTags),
 		},
-		Zones: &s.FailureDomains,
+		Zones: s.FailureDomains, // s.FailureDomains needs to be updated to a []*string
```
Comment (nawazkh) [inline azure/services/publicips/spec.go:100]: `// s.FailureDomains needs to be updated to a []*string`
This is more of an informational comment while removing. Will remove after review.

**Comment by nawazkh** [comment]
/hold
Merge https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/3969 before merging this PR

File: azure/converters/extendedlocation.go [hunk 2]
```diff
@@ -36,13 +35,13 @@ func ExtendedLocationToNetworkSDKv2(src *infrav1.ExtendedLocationSpec) *armnetwo
 }
 
 // ExtendedLocationToNetworkSDK converts infrav1.ExtendedLocationSpec to network.ExtendedLocation.
-func ExtendedLocationToNetworkSDK(src *infrav1.ExtendedLocationSpec) *network.ExtendedLocation {
+func ExtendedLocationToNetworkSDK(src *infrav1.ExtendedLocationSpec) *armnetwork.ExtendedLocation {
```
Comment (nawazkh) [inline azure/converters/extendedlocation.go:38]: We no longer need this function, will delete it.

Comment (nawazkh) [inline azure/converters/extendedlocation.go:38]: I dont really need to touch this function. Undoing the change.

File: azure/converters/extendedlocation_test.go [hunk 3]
```diff
@@ -63,17 +62,17 @@ func TestExtendedLocationToNetworkSDK(t *testing.T) {
 	tests := []struct {
 		name string
 		args *infrav1.ExtendedLocationSpec
-		want *network.ExtendedLocation
+		want *armnetwork.ExtendedLocation
```
Comment (nawazkh) [inline azure/converters/extendedlocation_test.go:65]: I don't need to update this test either in this PR. Will follow up in another PR.

File: azure/services/publicips/spec.go [hunk 4]
```diff
@@ -59,44 +59,44 @@ func (s *PublicIPSpec) OwnerResourceName() string {
...
-			PublicIPAllocationMethod: network.IPAllocationMethodStatic,
+		ExtendedLocation: converters.ExtendedLocationToNetworkSDKv2(s.ExtendedLocation),
+		Properties: &armnetwork.PublicIPAddressPropertiesFormat{
+			PublicIPAddressVersion:   &addressVersion,
+			PublicIPAllocationMethod: ptr.To(armnetwork.IPAllocationMethodStatic),
 			DNSSettings:              dnsSettings,
 			IPTags:                   converters.IPTagsToSDK(s.IPTags),
 		},
-		Zones: &s.FailureDomains,
+		Zones: s.FailureDomains, // s.FailureDomains needs to be updated to a []*string. This lead to updating ClusterDescriber interface's Zones to be []*string type
```
Comment (nawazkh) [inline azure/services/publicips/spec.go:100]: This is the reason for touching so many mocks.

Comment (nawazkh) [reply azure/services/publicips/spec.go:100]: I can instead add a converter here that converts `[]string` to `[]*string` and leave all the mocks as is..

File: azure/services/publicips/client.go [hunk 5]
```diff
@@ -18,139 +18,104 @@ package publicips
...
+	"sigs.k8s.io/cluster-api-provider-azure/azure/services/asyncpoller"
 	"sigs.k8s.io/cluster-api-provider-azure/util/reconciler"
 	"sigs.k8s.io/cluster-api-provider-azure/util/tele"
 )
 
 // AzureClient contains the Azure go-sdk Client.
 type AzureClient struct {
-	publicips network.PublicIPAddressesClient
+	publicips *armnetwork.PublicIPAddressesClient
+	auth      azure.Authorizer
```
Comment (nojnhuh) [inline azure/services/publicips/client.go:34]: It looks like `auth` isn't used.

File: azure/services/publicips/client.go [hunk 6]
```diff
@@ -18,139 +18,104 @@ package publicips
...
+func (ac *AzureClient) CreateOrUpdateAsync(ctx context.Context, spec azure.ResourceSpecGetter, resumeToken string, parameters interface{}) (result interface{}, poller *runtime.Poller[armnetwork.PublicIPAddressesClientCreateOrUpdateResponse], err error) {
 	ctx, _, done := tele.StartSpanWithLogger(ctx, "publicips.AzureClient.CreateOrUpdate")
 	defer done()
 
-	publicip, ok := parameters.(network.PublicIPAddress)
-	if !ok {
-		return nil, nil, errors.Errorf("%T is not a network.PublicIPAddress", parameters)
+	publicip, ok := parameters.(armnetwork.PublicIPAddress)
+	if !ok && parameters != nil {
+		return nil, nil, errors.Errorf("%T is not a armnetwork.PublicIPAddress", parameters)
```
Comment (nojnhuh) [inline azure/services/publicips/client.go:72]:
```suggestion
		return nil, nil, errors.Errorf("%T is not an armnetwork.PublicIPAddress", parameters)
```

File: azure/services/publicips/spec.go [hunk 7]
```diff
@@ -59,44 +59,44 @@ func (s *PublicIPSpec) OwnerResourceName() string {
 // Parameters returns the parameters for the public IP.
 func (s *PublicIPSpec) Parameters(ctx context.Context, existing interface{}) (params interface{}, err error) {
 	if existing != nil {
-		if _, ok := existing.(network.PublicIPAddress); !ok {
-			return nil, errors.Errorf("%T is not a network.PublicIPAddress", existing)
+		if _, ok := existing.(armnetwork.PublicIPAddress); !ok {
+			return nil, errors.Errorf("%T is not a armnetwork.PublicIPAddress", existing)
```
Comment (nojnhuh) [inline azure/services/publicips/spec.go:63]:
```suggestion
			return nil, errors.Errorf("%T is not an armnetwork.PublicIPAddress", existing)
```

File: azure/services/publicips/spec.go [hunk 8]
```diff
@@ -59,44 +59,44 @@ func (s *PublicIPSpec) OwnerResourceName() string {
...
 		})),
-		Sku:              &network.PublicIPAddressSku{Name: network.PublicIPAddressSkuNameStandard},
+		SKU:              &armnetwork.PublicIPAddressSKU{Name: ptr.To(armnetwork.PublicIPAddressSKUNameStandard)},
 		Name:             ptr.To(s.Name),
 		Location:         ptr.To(s.Location),
-		ExtendedLocation: converters.ExtendedLocationToNetworkSDK(s.ExtendedLocation),
-		PublicIPAddressPropertiesFormat: &network.PublicIPAddressPropertiesFormat{
-			PublicIPAddressVersion:   addressVersion,
-			PublicIPAllocationMethod: network.IPAllocationMethodStatic,
+		ExtendedLocation: converters.ExtendedLocationToNetworkSDKv2(s.ExtendedLocation),
```
Comment (nojnhuh) [inline azure/services/publicips/spec.go:93]: It looks like this was the last remaining instance of the "v1" version of this function, so we can probably remove that now and rename the v2 one to drop the "v2"

File: azure/services/publicips/client.go (hunk 5 above)
Comment (nawazkh) [reply azure/services/publicips/client.go]: Updated 7ba1619f6df44b7cc1ffb572c216a18297a7c4ed.

File: azure/services/publicips/client.go (hunk 6 above)
Comment (nawazkh) [reply azure/services/publicips/client.go]: Updated via 7ba1619f6df44b7cc1ffb572c216a18297a7c4ed

File: azure/services/publicips/spec.go (hunk 7 above)
Comment (nawazkh) [reply azure/services/publicips/spec.go]: Updated via 7ba1619f6df44b7cc1ffb572c216a18297a7c4ed

File: azure/services/publicips/spec.go (hunk 8 above)
Comment (nawazkh) [reply azure/services/publicips/spec.go]: Updated via 7ba1619f6df44b7cc1ffb572c216a18297a7c4ed

**Review by nojnhuh — COMMENTED** [review]
/lgtm

File: azure/services/publicips/client.go [hunk 9]
```diff
@@ -18,139 +18,103 @@ package publicips
...
+	}
 
-// newPublicIPAddressesClient creates a new public IP client from subscription ID.
-func newPublicIPAddressesClient(subscriptionID string, baseURI string, authorizer autorest.Authorizer) network.PublicIPAddressesClient {
-	publicIPsClient := network.NewPublicIPAddressesClientWithBaseURI(baseURI, subscriptionID)
-	azure.SetAutoRestClientDefaults(&publicIPsClient.Client, authorizer)
-	return publicIPsClient
+	factory, err := armnetwork.NewClientFactory(auth.SubscriptionID(), auth.Token(), opts)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create public IP client factory")
```
Comment (mboersma) [inline azure/services/publicips/client.go:45]:
```suggestion
		return nil, errors.Wrap(err, "failed to create armnetwork client factory")
```

**Review by mboersma — COMMENTED** [review]
This looks great! I only found two very nitpicky string changes for consistency with other services.

File: azure/services/publicips/client.go [hunk 10]
```diff
@@ -18,139 +18,103 @@ package publicips
...
-// NewClient creates a new public IP client from subscription ID.
-func NewClient(auth azure.Authorizer) *AzureClient {
-	c := newPublicIPAddressesClient(auth.SubscriptionID(), auth.BaseURI(), auth.Authorizer())
-	return &AzureClient{c}
-}
+// NewClient creates a new public IP client from an authorizer.
+func NewClient(auth azure.Authorizer) (*AzureClient, error) {
+	opts, err := azure.ARMClientOptions(auth.CloudEnvironment())
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create public IP client options")
```
[the remark this answers is not in the document]
Comment (nawazkh) [reply azure/services/publicips/client.go]: Updated via bc9a6ceb308c29b36642dc2479e31de195a761d9

File: azure/services/publicips/client.go (hunk 9 above)
Comment (nawazkh) [reply azure/services/publicips/client.go]: Updated via bc9a6ceb308c29b36642dc2479e31de195a761d9

**Review by mboersma — APPROVED** [review]
/lgtm
/approve

**Pull request merged**